### ISSUE TITLE
Kernel outlines don't contain enough const info for modular error checking / compilation

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/sdk#36635]
DDK and modular error checking assumes that kernel outlines contain full const info.  This allows tools to handle consts at a modular level without needing the whole program or being on the build critical path.  (This is how analyzer summaries / DDC works.)

Currently, kernel outlines don't appear to distinguish between a missing constant (?) and a `null` constant, so DDK builds will trigger unexpected `null` errors at runtime today.

Possibilities for 2.3:
- Use full kernel files for dependencies instead for now.
- Drop the const-update from 2.3.
- Fix issues around consts in outline.

We're exploring the first for now.  If it's not a significant perf regression, it's the easiest path.

**Comment by vsmenon** [comment]
Assigning to @jakemac53 to investigate the first option and perf implications.

**Comment by vsmenon** [comment]
fyi - @kmillikin @askeksa-google @stefantsov @jmesserly @natebosch @kevmoo

**Comment by vsmenon** [comment]
This is my update to Nate's CL (to use full kernel files):

https://dart-review.googlesource.com/c/sdk/+/99342

**Comment by jakemac53** [comment]
It looks like using full kernel files has pretty significant regression for full builds (about 2x worse on the gallery/spinning square), and a noticeable regression for incremental builds (~30% worse).

**Comment by vsmenon** [comment]
Okay, update on this: we'll likely roll forward with full kernel files.

The regression Jake reports roughly undoes the improvement we saw moving from analyzer to kernel.  We wouldn't block 2.3 on that, but I'd love to get the win back soon.

We're doing more validation.  If all is well, we'll aim to land a CL end of today and remove the P1 (and perhaps milestone).

**Comment by vsmenon** [comment]
@natebosch will work with @jakemac53 to land the logic for using full kernel files.

**Comment by natebosch** [comment]
Getting stuck. When we hacked in a hardcoded `summaryOnly = false` for the DDC path we could get things working, but being backwards compatible with the old path is tricky. The difference is the `--input-summary` vs `--input-linked` arguments. With DDC it only works with `--input-summary`.

**Comment by vsmenon** [comment]
Would it make sense to default / only support `summary-only` as `false` for now to unblock?

The summary path won't work properly yet anyway.

**Comment by natebosch** [comment]
Possibly. We already only support certain combinations of args and I'm worried about making that situation worse - in this scenario we only support using the arg `--summary-only`, but give an output like it was `--no-summary-only`...

**Comment by jmesserly** [comment]
Presumably, we'll also want to switch DDC's dart_sdk.dill file to use the full summary?

**Comment by natebosch** [comment]
https://dart-review.googlesource.com/c/sdk/+/99274 now includes a change to address the problem described in https://github.com/dart-lang/build/issues/2174 and I think it's both backwards compatible with existing usage of the kernel worker, and supports the enhanced collection literals assuming that the build system starts passing `--target` and `--no-summary-only`.

**Comment by jmesserly** [comment]
FYI, I'm testing https://dart-review.googlesource.com/c/sdk/+/99370, which switches DDC's "dartdevk" configuration to test the full Kernel IL file pipeline again. I'll also make sure that we build test package IL files (e.g. unittest, expect, etc) as full Kernel files.

**Comment by vsmenon** [comment]
@askeksa-google per @kmillikin , you're looking at this, so assigning to you.  feel free to push back to us once there is something to try out.

**Comment by vsmenon** [comment]
@kmillikin - here's a repro.  Sorry it's a bit messy.  We need to enable threading of experiments in package:build.

Create:
`const1.dart`:
```dart
const data1 = [1, 2, 3, if (false) 3.1413, 4, 5];
```

`const2.dart`
```dart
import 'const1.dart';

const data2 = {
  'data': [0, ...data1, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
};

void main() {
  print(data2);
  print(data2.length);
}
```

Run (you'll need to fix the paths):
```
> dartdevc -k --enable-experiment=constant-update-2018 --summary const1.dill --modules=node -o const2.js const2.dart
> dart ~/dart/sdk/utils/bazel/kernel_worker.dart --dart-sdk-summary=/Users/vsm/dart/sdk/xcodebuild/ReleaseX64/dart-sdk/lib/_internal/ddc_sdk.dill --packages-file=.packages --target ddc --source org-dartlang-app:/const1.dart --output const1.dill  --exclude-non-sources --multi-root-scheme org-dartlang-app 
> dartdevc -k --enable-experiment=constant-update-2018 --summary const1.dill --modules=node -o const2.js const2.dart
> NODE_PATH=~/dart/sdk/xcodebuild/ReleaseX64/dart-sdk/lib/dev_compiler/common/:. node -e 'require("./const2").const2.main()'

{data: [0, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]}
1
```

You can also just examine `const2.js`.  `data1` is null in the outline and inlined as such - i.e., the output is missing `1..5`.

**Comment by kmillikin** [comment]
Thanks for the test case, it definitely highlights the problem.  Serializing unevaluated constants in outlines doesn't "just" work because unevaluated constants can contain compiler-internal syntax that is not part of the Kernel language and impossible to serialize.

A solution is to compile that syntax away before serializing outlines (and without performing constant evaluation).  This issue is blocked on that task.

**Comment by kmillikin** [comment]
A proposed fix is at https://dart-review.googlesource.com/c/sdk/+/104564

**Comment by dgrove** [comment]
https://dart-review.googlesource.com/c/sdk/+/104564 has landed . can this be closed?

**Comment by kmillikin** [comment]
We need to verify that DDC is using Kernel outlines (I've lost track of whether it is or isn't) and whether constant-update-2018 works fully.

**Comment by vsmenon** [comment]
Filed a separate issue to validate and add a regression test:

https://github.com/dart-lang/sdk/issues/37161 

That only needs to block turning on the flag, not D24.

**Comment by kmillikin** [comment]
We might need more work, but the bulk of what we need is done.  Moving out of D24.

**Comment by vsmenon** [comment]
Yes, DDC is using kernel outlines, and this is what all external devs are using today.   The concern here is to make sure everything works with the const-flag on.

I will add a few regression tests and update this bug.

**Comment by vsmenon** [comment]
@kmillikin @askeksa-google 

I'm trying Flutter Web Gallery with the const flag enabled and hitting:
```
#12     ProgramCompiler.visitUnevaluatedConstant (package:dev_compiler/src/kernel/compiler.dart:5370:7)
#13     UnevaluatedConstant.accept (package:kernel/ast.dart:5727:34)
#14     ProgramCompiler.visitConstantExpression (package:dev_compiler/src/kernel/compiler.dart:3783:21)
```

Should we getting an unevaluated constant in the backend?

**Comment by askeksa-google** [comment]
An unavailable constructor turns into an unevaluated constant. See https://github.com/dart-lang/sdk/blob/f889bd70674248d3e24b99dd2ad25efdcf6bfee9/pkg/front_end/lib/src/fasta/kernel/constant_evaluator.dart#L873-L884

This is an indication that const constructors are still missing from the outlines.

Maybe the constant evaluator should not be lenient this way about unavailable constructors and variables. This code was adapted from DDC, but if we actually require these things to be available in outlines, it would make more sense for the constant evaluator to produce an internal error in this case (i.e. just not have the concept of unavailability).

**Comment by vsmenon** [comment]
Current repro steps:
- Enable flag (https://dart-review.googlesource.com/c/sdk/+/106943) and build SDK.
- Checkout flutter_web: https://github.com/flutter/flutter_web
- `cd flutter_web/examples/gallery`
- `pub get`
- `pub global activate webdev`
- Use built SDK to run, e.g.,: `PATH=~/dart/sdk/xcodebuild/ReleaseX64/dart-sdk/bin:$PATH webdev serve`

**Comment by kmillikin** [comment]
Thanks for the reproduction steps.  It turns out that as of https://github.com/dart-lang/sdk/commit/b32d196f00debeb38300d1a89c120958ceeda214 the SDK will no longer build with the flag enabled (filed https://github.com/dart-lang/sdk/issues/37357).  That commit doesn't revert cleanly so the simplest thing to do is to base the reproduction on the parent commit https://github.com/dart-lang/sdk/commit/b32d196f00debeb38300d1a89c120958ceeda214.

**Comment by mraleph** [comment]
This should now be fixed.
